crates/sui-graphql-rpc/src/types/transaction_block/tx_lookups.rs

```diff
@@ -0,0 +1,468 @@
+        })?;
+
+        // Expect exactly two rows, returning early if not.
+        let [(Some(db_lo_cp), Some(lo)), (Some(db_hi_cp), Some(hi))] = from_db.as_slice() else {
+            return Ok(None);
+        };
+
+        if *db_lo_cp as u64 != lo_cp.saturating_sub(1) || *db_hi_cp as u64 != hi_cp {
+            return Ok(None);
+        }
+
+        let lo = if lo_cp == 0 { 0 } else { *lo as u64 };
+        let hi = *hi as u64;
+
+        if page.after().is_some_and(|x| x.tx_sequence_number >= hi)
+            || page.before().is_some_and(|x| x.tx_sequence_number <= lo)
+        {
+            return Ok(None);
+        }
+
+        Ok(Some(Self::new(
+            lo,
+            hi,
+            page.after().map(|x| x.tx_sequence_number),
+            page.before().map(|x| x.tx_sequence_number),
+            scan_limit,
+            page.is_from_front(),
+        )))
+    }
+
+    pub(crate) fn tx_lo(&self) -> u64 {
+        max_option([self.after, Some(self.lo)]).unwrap()
+    }
+
+    pub(crate) fn tx_hi(&self) -> u64 {
+        min_option([self.before, Some(self.hi)]).unwrap()
+    }
+
+    /// The lower bound `tx_sequence_number`` of the range to scan within. This defaults to the min
+    /// tx_sequence_number of the checkpoint bound. If a cursor is provided, the lower bound is
+    /// adjusted to the larger of the two. The resulting value is additionally modified by the
+    /// `scan_limit` if `is_from_front` is false.
+    pub(crate) fn scan_lo(&self) -> u64 {
+        let adjusted_lo = self.tx_lo();
+
+        if self.is_from_front {
+            adjusted_lo
+        } else {
+            // If not from the front, then the scan_limit must only be applied to the lower bound
+            if let Some(scan_limit) = self.scan_limit {
+                adjusted_lo.max(self.tx_hi().saturating_sub(scan_limit))
+            } else {
+                adjusted_lo
+            }
+        }
+    }
+
+    /// The upper bound `tx_sequence_number` of the range to scan within. This defaults to the max
+    /// tx_sequence_number of the checkpoint bound. If a cursor is provided, the upper bound is
+    /// adjusted to the smaller of the two. The resulting value is additionally modified by the
+    /// `scan_limit` if `is_from_front` is true.
+    pub(crate) fn scan_hi(&self) -> u64 {
+        let adjusted_hi = self.tx_hi();
+
+        if self.is_from_front {
+            if let Some(scan_limit) = self.scan_limit {
+                adjusted_hi.min(self.tx_lo().saturating_add(scan_limit))
+            } else {
+                adjusted_hi
+            }
+        } else {
+            adjusted_hi
+        }
+    }
+
+    /// If the query result does not have a previous page, check whether the scan limit is within
+    /// the initial tx_sequence_number range.
+    pub(crate) fn scan_has_prev_page(&self) -> bool {
+        if self.after.unwrap_or(0) >= self.hi {
+            return false;
+        }
+
+        self.scan_lo() > self.lo
```

Comment:
Trying to wrap my head around this first condition and how it maps to the query not having a previous page. We're saying the overall condition is:

```rust
self.after.unwrap_or(0) < self.hi && self.scan_lo() > self.lo
```

`self.after` is the lowerbound from the cursor, and `self.hi` is the upperbound from the transaction sequence number filters, so I'm not seeing how this comparison demonstrates that there isn't a previous page. For that I would expect something like: `self.after.unwrap_or(0) <= self.lo` (if the cursor is at or below the lowerbound then we don't have a previous page), so overall I would expect this condition to be:

```rust
self.after.unwrap_or(0) <= self.lo && self.lo < self.scan_lo()
```

(Also note how the sanity preserving trick of keeping all inequalities monotonic helps make things clear here).

Reply:
This is indeed fairly confusing. I have a more n00b question though: wouldn't the existence of `self.after` imply that a prev page exists? Since this `after` came from the page cursor, which, I suppose, came from when the previous page was returned? Under what scenario would we have a `Some` `after` value that's smaller than `self.lo`? Maybe I'm misunderstanding the intention of this `scan_has_prev_page` altogether.

Reply:
we can drop the part that isn't `self.lo < self.scan_lo()`, that was a vestige left behind from some other bit 

> Under what scenario would we have a Some after value that's smaller than self.lo

Yeah totally, I ... forgot why I even had this condition here. Seems fine to remove. Returning to your earlier point/ question: I think we can't assume that an `after` from the caller will guarantee that a previous page exists. The current implementation of `paginate_results` explicitly checks this, for example if paginating forwards, then the left tip's cursor value should equal `after`, and if not, then we return an empty connection.

Now that we have `scan_limit`ed cursors, the above is a little too strict, and we need to loosen it up a little. So if the `after` cursor provided comes from a `scan_limit`, the check is loosened to "there is a previous page as long as the given cursor is not at the lower bound of the tx range"